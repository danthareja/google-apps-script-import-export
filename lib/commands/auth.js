const colors = require('colors');
const readline = require('readline');
const OAuth2Client = require('googleapis').auth.OAuth2;

const Promise = require('bluebird');
const fs = Promise.promisifyAll(require('fs'), { multiArgs: true });

const defaults = require('../defaults');

const GOOGLE_AUTH_SCOPE = [
  'https://www.googleapis.com/auth/drive',
  'https://www.googleapis.com/auth/drive.scripts'
];

/**
 * Tries to find the storage file. If found, tells the users to remove it, otherwiser runs the
 * authentication flow using the provided path the to the clientSecret path
 * @param {String} clientSecretPath
 * @param {Boolean} withWebServer
 * @returns {Promise}
 */
module.exports = function auth(clientSecretPath, withWebServer = true) {
  return fs
    .lstatAsync(defaults.STORAGE_FILE)
    .then(() => {
      console.log(
        defaults.STORAGE_FILE +
          ' already exists. Remove it to re-authenticate node-google-apps-script'
      );
      return Promise.resolve();
    })
    .catch(err => {
      if (!clientSecretPath) {
        throw defaults.STORAGE_FILE +
          ' does not exist yet. Specify a credential file with `--auth ~/Downloads/client_secret_abcd.json`';
      }
      return performAuthenticationFlow(clientSecretPath, withWebServer);
    });
};

/**
 * Performs the authentication flow by reading the .json file
 * and using it's content to authenticate with Google. When
 * done, clean ups the stdin.
 * @param {String} clientSecretPath
 * @param {Boolean} withWebServer
 * @return {Promise}
 */
function performAuthenticationFlow(clientSecretPath, withWebServer) {
  return getCredentialsFromFile(clientSecretPath)
    .then(credentials => authenticateWithGoogle(credentials, withWebServer))
    .then(saveAuthenticationConfig)
    .then(cleanUp)
    .then(() => {
      console.log('Successfully Authenticated with Google Drive!'.green);
    })
    .catch(err => {
      console.log('Error running auth command'.underline.red);
      console.log(err.red);
      throw err;
    });
}

/**
 * The parsed credentials
 * @typedef {Object} Credentials
 * @property {String} client_id
 * @property {String} client_secret
 * @property {String} redirect_uri
 * @property {String} [refresh_token]
 */
/**
 * Reads the content of the credentials file and returns a object
 * with the important fields
 * @param {String} clientSecretPath
 * @returns {Promise<Credentials>}
 */
function getCredentialsFromFile(clientSecretPath) {
  return fs
    .readFileAsync(clientSecretPath)
    .then(JSON.parse)
    .then(credentials => {
      if (!credentials.installed) {
        throw 'Path did not include "OAuth 2.0 client ID" credentials. Please check that you downloaded the right JSON credentials.';
      }

      // Add important auth information to persist
      // return credentials
      return {
        client_id: credentials.installed.client_id,
        client_secret: credentials.installed.client_secret,
        redirect_uri: credentials.installed.redirect_uris[0]
      };
    })
    .catch(SyntaxError, err => {
      console.log('Invalid file contents'.red);
    })
    .error(err => {
      console.log('Unable to read file'.red);
      throw err;
    });
}

/**
 * Authenticate with Google using the parsed credentials.
 * Uses the OAuth2Client to generate the AuthUrl
 * @param {Credentials} credentials
 * @param {Boolean} withWebServer
 * @returns {Promise}
 */
function authenticateWithGoogle(credentials, withWebServer) {
  return new Promise((resolve, reject) => {
    let redirect_uri = credentials.redirect_uri;
    if (withWebServer) {
      redirect_uri = 'http://localhost:' + defaults.WEBSERVER_PORT;
    }

    const oauth2Client = new OAuth2Client(
      credentials.client_id,
      credentials.client_secret,
      redirect_uri
    );
    const url = oauth2Client.generateAuthUrl({
      access_type: 'offline', // Required to receive a refresh token
      approval_prompt: 'force',
      scope: GOOGLE_AUTH_SCOPE
    });

    const getToken = code => {
      oauth2Client.getToken(code, (err, tokens) => {
        if (err) {
          reject(err);
          return;
        }
        credentials.refresh_token = tokens.refresh_token;
        resolve(credentials);
      });
    };

    console.log(
      "\nPlease visit the following url in your browser (you'll only have to do this once):"
        .cyan,
      url.green
    );

    if (withWebServer) {
      require('../miniOAuthServer')(defaults.WEBSERVER_PORT)
        .then(getToken)
        .catch(err => {
          console.log('Something went wrong with the webserver'.red);
          throw err;
        });
    } else {
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
      });

      rl.question('Copy the provided code and paste it here: ', code => {
        rl.close();

        getToken(code);
      });
    }
  });
}

/**
 * Write the contents to disk
 * @param {Credentials} credentials
 */
function saveAuthenticationConfig(credentials) {
  return fs.writeFileAsync(
    defaults.STORAGE_FILE,
    JSON.stringify(credentials, '', 2)
  );
}

/**
 * Destroys the stdin
 * @todo put inline if only used here
 */
function cleanUp() {
  process.stdin.destroy();
}
