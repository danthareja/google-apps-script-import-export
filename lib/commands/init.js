const colors = require('colors');
const mkdirp = require('mkdirp');

const Promise = require('bluebird');
const fs = Promise.promisifyAll(require('fs'));

const util = require('../util')
const defaults = require('../defaults');
const manifestor = require('../manifestor');

module.exports = function init(fileId, options) {
  const subdir = options.subdir || defaults.DEFAULT_SUBDIR;
  
  if(!fileIdIsValid(fileId)) {
    return;
  }
  
  const config = {
    path: subdir,
    fileId: fileId,
    key: options.key
  };

  const overwritePromise = options.overwrite ?
    Promise.resolve() :
    manifestor.throwIfConfig();

  return overwritePromise
    .then(function() {
      return manifestor.set(config);
    })
    .then(function() {
      return mkdirp(subdir);
    })
    .then(function(config) {
      return manifestor.getExternalFiles(fileId)
    })
    .map(function(file) {
      return writeExternalFile(file, subdir)
    })
    .catch(function(err) {
      console.log('Error running init command'.red);
      throw err;
    });
};

function writeExternalFile(file, dir) {
  const filename = file.name + util.getFileExtension(file)
  return fs.writeFileAsync(dir + '/' + filename, file.source)
    .catch(function(err) {
      console.log('Could not write file ' + filename);
      throw err;
    })
}

function fileIdIsValid(fileId) {
  if(fileId.charAt(0).toLowerCase() === 'm') {
    console.log('\nScript file ID error.'.red + '\n' + 
        'It looks like you are passing in a Project Key, from "File --> Project properties",' +
        'rather than a Drive File ID.\nYou will find the Drive File ID in the script\'s URL:\n' +
        'https://script.google.com/a/google.com/d/' + '__DRIVE_FILE_ID__'.green + '/edit.\n');
     return false;
  } else {
     return true;
  } 
}