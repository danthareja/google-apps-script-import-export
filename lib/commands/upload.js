const _ = require('lodash');
const colors = require('colors');
const google = require('googleapis');
const Promise = require('bluebird');

const defaults = require('../defaults');
const manifestor = require('../manifestor');
const authenticate = require('../authenticate');

module.exports = function upload() {
  console.log('Pushing back up to Google Drive...');

  let fileId; // Hold in closure to avoid promise nesting

  return manifestor.get()
    .then((config) => {
      fileId = config.fileId;
      return manifestor.getExternalFiles(fileId)
    })
    .then((externalFiles) => manifestor.build(externalFiles))
    .then((files) => sendToGoogle(files, fileId))
    .then(() => {
      console.log('The latest files were successfully uploaded to your Apps Script project.'.green);
    })
    .catch((err) => {
      console.log('Upload failed.'.red);
    });
};

function sendToGoogle(files, id) {
  if (!files.length) {
    console.log('No Files to upload.'.red);
    throw 'manifest file length is 0';
  }

  return authenticate()
    .then((auth) => {
      const drive = google.drive({ version: 'v2', auth: auth });
      const options = {
        fileId: id,
        media: {
          mimeType: 'application/vnd.google-apps.script+json',
          body: JSON.stringify({ files: files })
        }
      };

      return Promise.promisify(drive.files.update)(options)
        .catch((err) => {
          console.log('An error occured while running upload command: '.red + err.message);
          throw err;
        });
    });
}
