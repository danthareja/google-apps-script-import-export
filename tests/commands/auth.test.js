const fs = require("fs");
const { promisify } = require("util");
const unlinkAsync = promisify(fs.unlink);
jest.mock("../../lib/defaults");

const fakeClientSecretPath = "/tmp/.fakeClientSecret";
const defaults = require("../../lib/defaults");

const removeFile = () => unlinkAsync(defaults.STORAGE_FILE).catch(() => {});
const writeFakeFile = () => fs.writeFileSync(defaults.STORAGE_FILE, "f4k3");

let spiedConsole = jest.spyOn(console, "log").mockImplementation(() => {});

const cleanUp = () => {
  jest.resetAllMocks();
  removeFile();
};

const clientSecret = {
  installed: {
    client_id: "424242424242-f4k3h4ck15h.apps.googleusercontent.com",
    project_id: "project-id-424242424242",
    auth_uri: "https://accounts.google.com/o/oauth2/auth",
    token_uri: "https://accounts.google.com/o/oauth2/token",
    auth_provider_x509_cert_url: "https://www.googleapis.com/oauth2/v1/certs",
    client_secret: "client_secret_fake",
    redirect_uris: ["urn:ietf:wg:oauth:2.0:oob", "http://localhost"]
  }
};


afterEach(cleanUp);

test("should resolve the promise when a file already exists", async () => {
  writeFakeFile();

  const auth = await require("../../lib/commands/auth")();

  expect(auth).toBeUndefined();
  expect(spiedConsole.mock.calls[0][0]).toMatchSnapshot();
});

test("should reject the promise when a previous file does not exists and a credential path is not provided", async () => {
  try {
    await require("../../lib/commands/auth")();
  } catch (e) {
    expect(e).toMatchSnapshot();
  }
});

test("should start a authentication flow and print the oauth url when a clientSecret is provided", async done => {
  fs.writeFileSync(fakeClientSecretPath, JSON.stringify(clientSecret), "utf-8");

  require("../../lib/commands/auth")(fakeClientSecretPath);

  // TODO: Find a better way
  // Now we are using setTimeout just to have enought time for the async code to run
  setTimeout(() => {
    expect(spiedConsole.mock.calls[0][0]).toMatchSnapshot();
    expect(spiedConsole.mock.calls[0][1]).toContain("https://");
    done();
  }, 1000);
});

test("should complain if credentials are not in the OAuth 2.0 format", async done => {
  const wrongFormatSecret = {
    client_id: "424242424242-f4k3h4ck15h.apps.googleusercontent.com",
    project_id: "project-id-424242424242"
  };
  fs.writeFileSync(
    fakeClientSecretPath,
    JSON.stringify(wrongFormatSecret),
    "utf-8"
  );

  try {
    await require("../../lib/commands/auth")(fakeClientSecretPath);
  } catch (e) {
    expect(e).toMatchSnapshot();
    done();
  }
});
