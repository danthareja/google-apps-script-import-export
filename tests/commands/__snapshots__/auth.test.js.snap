// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should complain if credentials are not in the OAuth 2.0 format 1`] = `"Path did not include \\"OAuth 2.0 client ID\\" credentials. Please check that you downloaded the right JSON credentials."`;

exports[`should reject the promise when a previous file does not exists and a credential path is not provided 1`] = `"/tmp/.gapps-test does not exist yet. Specify a credential file with \`--auth ~/Downloads/client_secret_abcd.json\`"`;

exports[`should resolve the promise when a file already exists 1`] = `"/tmp/.gapps-test already exists. Remove it to re-authenticate node-google-apps-script"`;

exports[`should start a authentication flow and print the oauth url when a clientSecret is provided 1`] = `
"[36m
Please visit the following url in your browser (you'll only have to do this once):[39m"
`;
